### DISCUSSION TITLE
Chainloading from pxelinux (DVD boot) to ipxe on HP DL380 Gen 9

### DISCUSSION THREAD

**Discussion by vtwaldo21** [ipxe/ipxe#1196] [Q&A] [answered]
I apologize as I really don't know where to begin. I've been using ipxe for a while and been very happy. I have some "new" HW, the HP DL380 Gen 9 (latest bios) but having issues loading ipxe.lkrn from the syslinux menu. Basically, my make command embeds two files, an ipxe script and the vmlinuz kernel. e.g,

make bin/ipxe.lkrn EMBED=start.ipxe,vmlinuz

Then the isolinux.cfg file on the bootable DVD has the lines
KERNEL ipxe.lkrn
APPEND initrd=initrd.img

This exact setup how worked for years for me. But trying this method on this "new" hardware, causes the system to just reboot after the syslinux menu item is selected.  I can't actually find where this method is  documented anywhere, but I doubt I figured it out without online help years ago :)

The vmlinux file is ~ 11MB.  If I remove that from the EMBED line, it works (I get the IPXE menus), but of course I then have no kernel to load. I could pass it in via the initrd line, but then, i have no initrd image to load. I can also pass both the initrd and the vmlinux via, initrd=vmlinuz,initrd.img and ipxe gets the two files concatenated as one (via imgstat output).  Standard documentation would have you download the vmlinuz or initrd off the network but this is a purely from DVD install, the network is not available. 

Testing around, I could add other files to the EMBED but at some size point (which I have yet to determine), just causes the system to reset.  There is nothing on the screen anywhere and I've attempted to enable all the debugging I could think of but still nothing.

What I don't understand is the ipxe.lkrn file is actually smaller than the standalone vmlinuz file which I can use in syslinux and boot just fine.  As I said, this set up works for me on many other systems as well as VMs. So i'm sure it is a "HW" issue, but I dont' even understand what it could be. There are 100GB+ of RAM, but i'm guessing this is lower level in the boot rom which I dont' understand at all. 

Perhaps also, my method to EMBED the vmlinuz file is incorrect. But I can't find any examples that show how to boot from a DVD, use an ipxe script/menu system and then start linux. As best as I can tell, i have to have both vmlinux and initrd.img "passed" to ipxe because there is not filesystem access that I know of.

Any and all help is much appreciated, thank you.

**Comment by NiKiZe** [comment]
Why are you using iPXE, if you only want local boot anyway?

Have you considered moving to efi?

Does the exact same dvd work on different hardware? (Or virtual machines)?

**Reply by vtwaldo21** [reply]
Thanks for the reply!

 > Why are you using iPXE, if you only want local boot anyway?

I prefer the menu system just to be more user friendly for non technical people. But mostly because it also allows (with some minor code tweaks) graphical user input. Also, it then makes the installation the same if installing from local media or if in an environment that actually has a pxe server. This limits end user documentation.

> Have you considered moving to efi?

Considered it a little bit. Moving to EFI is probably not a realistic option only because this kind of program fears change.  But it is a good point and one I'll retest against at least as a data point.

 > Does the exact same dvd work on different hardware? (Or virtual machines)?

Yes, it works fine on every other system I've tried. Which granted is less than a dozen...

**Comment by vtwaldo21** [comment]
i'm not having any luck building a UEFI bootable iso to do the sanity check. The instructions I found have you just build ipxe.efi and then run genfsimg but KVM at least seems to not think it is a valid boot device (which I see other reports for as well online but very dated so probably not the same issue)

**Reply by vtwaldo21** [reply]
this was probably just operator error. I've now learned that to configure KVM to use UEFI you can't select the basic UEFI option or it doesn't seem to work. You have to select a specific UEFI image. I haven't gone back and tested the ipxe UEFI but it probably would work if KVM was configured properly.

**Comment by mcb30** [comment, marked as answer]
> The vmlinux file is ~ 11MB. If I remove that from the EMBED line, it works (I get the IPXE menus), but of course I then have no kernel to load. I could pass it in via the initrd line, but then, i have no initrd image to load. I can also pass both the initrd and the vmlinux via, initrd=vmlinuz,initrd.img and ipxe gets the two files concatenated as one (via imgstat output). Standard documentation would have you download the vmlinuz or initrd off the network but this is a purely from DVD install, the network is not available.
> 
> Testing around, I could add other files to the EMBED but at some size point (which I have yet to determine), just causes the system to reset. There is nothing on the screen anywhere and I've attempted to enable all the debugging I could think of but still nothing.

There are all sorts of hidden oddities in the BIOS system memory map.  Most likely something undocumented in the BIOS is using a block of memory at around ~16MB, which you end up overwriting when your .lkrn image gets loaded over the top of it.

iPXE deliberately relocates itself to the highest available address (below 4GB) as part of a strategy to avoid this kind of problem, since most of the undocumented BIOS uses seem to occur lower down in the memory map.  However, if you've already overwritten whatever was at 16MB, then subsequently relocating to a higher address won't help.

> What I don't understand is the ipxe.lkrn file is actually smaller than the standalone vmlinuz file which I can use in syslinux and boot just fine. As I said, this set up works for me on many other systems as well as VMs. So i'm sure it is a "HW" issue, but I dont' even understand what it could be. There are 100GB+ of RAM, but i'm guessing this is lower level in the boot rom which I dont' understand at all.

It could be that the undocumented area being overwritten happens during the self-decompression of iPXE.  This will (for an .lkrn image) use a temporary area somewhere below 64MB.

You could try hacking libprefix.S to change `'#define COMPRESS 1` to `#define COMPRESS 0`.  This should give you a larger iPXE image, but then not need to use a temporary decompression area.

>  I prefer the menu system just to be more user friendly for non technical people. But mostly because it also allows (with some minor code tweaks) graphical user input. Also, it then makes the installation the same if installing from local media or if in an environment that actually has a pxe server. This limits end user documentation.

It's nice to hear that you prefer the iPXE menu system.  :slightly_smiling_face: 

> Perhaps also, my method to EMBED the vmlinuz file is incorrect. But I can't find any examples that show how to boot from a DVD, use an ipxe script/menu system and then start linux. As best as I can tell, i have to have both vmlinux and initrd.img "passed" to ipxe because there is not filesystem access that I know of.

You are correct that there is no filesystem access and so you have to arrange for files to be passed to iPXE (either via embedding, or as an initrd to the .lkrn image).

There is support in iPXE for extracting files from an archive format.  At present this supports only gzip and zlib, both of which are single-file archive formats.  It could potentially be extended to handle multiple-file archives such as tar or zip: this would allow you to provide an archive containing both kernel+initrd as the "initrd" to the ipxe.lkrn file, and then extract the kernel and initrd separately in iPXE using `imgextract`.

**Comment by vtwaldo21** [comment]
Thanks for the details and tips. I tried setting compression to 0 but same issue. I had no idea about the imgextract command. Extending that to support tar or something would be clever indeed. I'll have a look at it.

Just for completeness, I also tested on a second HP DL380 just to mostly rule out complete random HW issue.

**Comment by vtwaldo21** [comment]
unrelated to my base ticket, but this code seems questionable to me,

hci/commands/image_archive_cmd.c:
```
   /* Extract archive image */
   if ( ( rc = imgextract ( image, opts.name ) ) != 0 )
      goto err_extract;

   /* Success */
   rc = 0;

err_extract:
   /* Discard original image unless --keep was specified */
   if ( ! opts.keep )
      unregister_image ( image );
 err_acquire:
 err_parse:
   return rc;
}
```

Wouldn't you want the label for err_extract to be after the unregister_image?

Likewise core/archive.c:
```
  /* Drop local reference to image */
   image_put ( *extracted );

   return 0;

   unregister_image ( *extracted );
 err_register:
 err_extract:
 err_set_name:
   image_put ( *extracted );
 err_alloc:
```

Seems like the unregister_image line will not be called.

**Reply by mcb30** [reply]
> ```
>   /* Drop local reference to image */
>    image_put ( *extracted );
> 
>    return 0;
> 
>    unregister_image ( *extracted );
>  err_register:
>  err_extract:
>  err_set_name:
>    image_put ( *extracted );
>  err_alloc:
> ```
> 
> Seems like the unregister_image line will not be called.

This is the general structured error handling approach used in iPXE.  The "undo" functions (such as `unregister_image()`) are designed to be called only after a **successful** call to the corresponding "do" functions (such as `register_image()`).  The general pattern is therefore:

```
    if ( ( rc = try_thing_one() ) != 0 ) {
        DBG ( "thing one failed: %s\n", strerror ( rc ) );
        goto err_thing_one;
    }
    ...
    if ( ( rc = try_thing_two() ) != 0 ) {
        DBG ( "thing two failed: %s\n", strerror ( rc ) );
        goto err_thing_two;
    }
    ...    

    ...
    undo_thing_two();
  err_thing_two:
    ...
    undo_thing_one();
  err_thing_one:
    ...
```
i.e. the undo function always appears directly **before** the relevant error label, and error labels appear in reverse order to their corresponding `goto` statements.

This pattern allows for systematic stacking of multiple error handling paths.  It also means that code diffs are clean: the change to add a new possible failure path (e.g. a memory allocation) never needs to modify any existing error handling code, it just needs to insert its own error handling code at the correct place within the existing error handling code.

**Reply by vtwaldo21** [reply]
yeah, i believe that was my point. Like in the first example, I read that as NOT following that paradigm:

Attempt to extract image.
If it fails, delete the original image.

I believe it should only delete the original image if the extraction was successful. Otherwise, you have deleted the original file and not produced the extracted version?

As for the second example, my understanding for that one is any line after the "return 0" and before any labels could never be executed. I dont' see an execution path that could lead to that command.

**Reply by mcb30** [reply]
> yeah, i believe that was my point. Like in the first example, I read that as NOT following that paradigm:
> 
> Attempt to extract image. If it fails, delete the original image.

Attempt to extract image.  Whether it succeeds or fails, delete the original (compressed) image, unless the user explicitly indicated that the original image was to be retained.

> As for the second example, my understanding for that one is any line after the "return 0" and before any labels could never be executed. I dont' see an execution path that could lead to that command.

There is no current execution code path that could lead to that statement.  But a future patch may add a code path that could lead to that statement, and I don't want that future patch to be forced to add in error handling code that is entirely unrelated to what that future patch is trying to achieve.

**Reply by vtwaldo21** [reply]
roger, thanks for the explanations.

**Comment by vtwaldo21** [comment]
@mcb30 , so your idea to implement a multi-file archive worked. I implemented an option to support the tar format (uncompressed) and was able to pass a tar file on the append line with multiple files embedded and they all appear by individual name when using imgstat and can then be passed to the kernel command as one would expect. That does seem like a pretty clean solution. 

Of course i just completely hacked in support for a tar file, a fair bit more work would be needed to do it right. But luckily the format can be interpreted with functions that are already part of ipxe, mostly just strtoul and copy_from_user. 

Thanks again for your help!

**Comment by keiser1080** [comment]
@vtwaldo21

could you share an example of your start.ipxe ?
I am trying to replace grub on ubuntu server iso by ipxe.
But i diden't found how to load from the iso fs the kernel and the initrd.
